### PR TITLE
Upgrade Flutter and packages; hold back firebase_core; record Flutter commit IDs

### LINKED ISSUE #1118: Record Flutter commit ID as well as version number
This issue can be seen as a much-scoped-down version of pinning the Flutter version:
* #15

That issue has stalled because the solutions to it tend to involve inserting themselves into the day-to-day development workflow — after all, they need to control what version of Flutter gets used — and none of the known solutions have felt totally satisfactory to me. Also because, in practice, our existing workflow without actual pinning generally works just fine.

There is one aspect that's bugged me a few times so far in practice, though, and which doesn't call for changing anything in the day-to-day workflow so it should be much easier to resolve: I'd like to know the Git commit ID for the Flutter version that we were using at a specific point in the repo's history. Specifically in practice, I want to know what Flutter commit ID was used for a given published (beta) release.

I already make a practice of building each release with the exact minimum version recorded in pubspec.yaml, in order to make it reproducible. So this information is there in principle, encoded in that version number. But from those Flutter version numbers (`3.27.0-1.0.pre.518`, in our `pubspec.yaml` in main at the moment) it's somewhat annoying to identify the actual Git commit ID (`01590aa27ac4d3a904d25ae4c6b792f933a3da6f` in that example, i.e. flutter/flutter@01590aa27ac4d3a904d25ae4c6b792f933a3da6f).

We can solve this by just having the commit ID also in pubspec.yaml, e.g. in a comment on the same line. Then `tools/upgrade` can maintain the comment. Ideally, we can even have a check in `tools/check` that verifies it's accurate.

### LINKED ISSUE #1119: Dart SDK change breaks background notifications
On upgrading to Flutter from current main, and running in release mode, the app no longer shows notifications on Android when it was in the background. This burst of error messages appears in logcat:

```
DartVM   E  ERROR: To access 'package:zulip/notifications/receive.dart::NotificationService' from native code, it must be annotated.
            ERROR: See https://github.com/dart-lang/sdk/blob/master/runtime/docs/compiler/aot/entry_point_pragma.md
flutter  E  [ERROR:flutter/lib/ui/dart_runtime_hooks.cc(38)] Dart Error: ERROR: To access 'package:zulip/notifications/receive.dart::NotificationService' from native code, it must be annotated.
            ERROR: See https://github.com/dart-lang/sdk/blob/master/runtime/docs/compiler/aot/entry_point_pragma.md
flutter  E  [ERROR:flutter/shell/common/shell.cc(115)] Dart Error: ERROR: To access 'package:zulip/notifications/receive.dart::NotificationService' from native code, it must be annotated.
            ERROR: See https://github.com/dart-lang/sdk/blob/master/runtime/docs/compiler/aot/entry_point_pragma.md
flutter  E  [ERROR:flutter/shell/common/shell.cc(115)] Dart Error: ERROR: To access 'package:zulip/notifications/receive.dart::NotificationService' from native code, it must be annotated.
                                                                                                    ERROR: See https://github.com/dart-lang/sdk/blob/master/runtime/docs/compiler/aot/entry_point_pragma.md
```

It looks like the relevant upstream commit is probably this one:
https://github.com/dart-lang/sdk/commit/9b06e26620a081c89931e5618336ff23dc6083be
which is also the latest commit to touch that linked docs page.

AFAICT those docs don't actually say that we should need to annotate the class. We already do annotate the static *method* that we ask native code to call:
```dart
  @pragma('vm:entry-point')
  static Future<void> _onBackgroundMessage(FirebaseRemoteMessage message) async {
```
Well, at least the error message is nice and clear, though. We'll annotate the class too.

### PR DESCRIPTION
**Pull request by gnprice**
Fixes #1118.
Fixes #1119.

This upgrades Flutter and some of our other dependencies.

OTOH `firebase_core` and the other Firebase packages can't immediately be upgraded; I ran into #1116, which I just filed. So hold those back.

And at the start of the branch, I augment `tools/upgrade` a bit so that it records the actual commit ID in the Flutter tree in a comment in pubspec.yaml, and give `tools/check` a tiny new suite `flutter_version` that validates that information.

On several occasions I've wanted the ability to look back and see what Flutter commit was used in a given published (beta) release, in particular. I already make a practice of doing so with the exact minimum version recorded in `pubspec.yaml`, in order to make it reproducible; but it's somewhat annoying to identify that commit from the version number that normally appears there. So this solves that problem. (… OK, and filed #1118 to describe that.)

# Selected commit messages

#### f5395d990 tools/upgrade: Record commit ID of Flutter into pubspec.yaml

Fixes #1118, or anyway the main part of it.

Upcoming commits will add a suite in tools/check that validates
this information, in case this line gets updated manually rather
than using this script.

#### a29b6143a notif: Add vm:entry-point pragma on class, too

Fixes #1119.
[…]

#### b1e77284e deps: Upgrade packages within constraints (tools/upgrade pub)

This required a manual fixup step to update generated files
for the Drift upgrade:
  $ tools/check --fix --all-files build_runner drift

### CODE REVIEW

**Comment by gnprice** [comment]
CI failed initially but should be fixed now. The fix was adding this commit:
9618f40aa check: Fix branch name for clone of Flutter repo in CI

File: tools/check
Comment (chrisbobbe) [inline tools/check:283]: I've named the remote "upstream" instead of "origin", so "origin/main" doesn't work for me (here and below). I've named it back to "origin" for testing this new code; should I keep it that way or should the code adapt to handle the "upstream" name?

**Review by chrisbobbe — APPROVED** [review]
Thanks! LGTM with one small question below.

**Comment by chrisbobbe** [comment]
Oh I did notice an auto-generated change:

```
diff --git ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme
index 5e31d3d34..c53e2b314 100644
--- ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme
+++ ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme
@@ -48,6 +48,7 @@
       ignoresPersistentStateOnLaunch = "NO"
       debugDocumentVersioning = "YES"
       debugServiceExtension = "internal"
+      enableGPUValidationMode = "1"
       allowLocationSimulation = "YES">
       <BuildableProductRunnable
          runnableDebuggingMode = "0">
```

File: tools/check
Comment (gnprice) [reply tools/check:283]: I'd suggest just switching back to "origin". It'd be possible to make this handle both conventions, but would make the logic more complicated.

I don't think the Flutter install steps do anything to override Git's default here, so most contributors will have "origin".

Comment (gnprice) [reply tools/check:283]: (Personally I switched back from using "upstream" to using "origin" a little while ago, for Git repos in general, partly just because I find "origin" is quicker for me to reliably type — "upstream" has long runs on the same hand.)

**Comment by chrisbobbe** [comment]
Hmm, seeing some iOS build issues actually—please hold off merging while I investigate. Here is the output:

```
Launching lib/main.dart on iPhone 15 Pro in debug mode...
Running pod install...
Running Xcode build...
Xcode build done.                                           30.5s
Failed to build iOS app
Semantic Issue (Xcode): Duplicate interface definition for class 'Sqlite3FlutterLibsPlugin'
/Users/chrisbobbe/dev/zulip-flutter/build/ios/Debug-iphonesimulator/sqlite3_flutter_libs/sqlite3_flutter_libs.framework/Headers/Sqlite3FlutterLibsPlugin.h:2:0

Modules Issue (Xcode): Definition of 'Sqlite3FlutterLibsPlugin' must be imported from module 'sqlite3_flutter_libs.Swift' before it is required
/Users/chrisbobbe/dev/zulip-flutter/ios/Runner/GeneratedPluginRegistrant.m:105:3

Could not build the application for the simulator.
Error launching application on iPhone 15 Pro.
```

**Comment by chrisbobbe** [comment]
Hmm, the build succeeded just now when I requested it from the Xcode gui. The output above is from `flutter run` as triggered from the "Run 'main.dart'" button (with a green "play" icon) in Android Studio.

**Comment by chrisbobbe** [comment]
> The output above is from flutter run as triggered from the "Run 'main.dart'" button (with a green "play" icon) in Android Studio.

OK, and it succeeds for me now after running `flutter clean`. 🤷‍♂️ Please merge at will, then.

**Comment by gnprice** [comment]
Thanks for the review! Evidently I had not tried `flutter run` for iOS. <_< >_>

I reproduce that build failure, and also reproduce that `flutter clean` resolves it — good find.

Before `flutter clean`, I tried `flutter build ipa`, prompted by thinking of #773. That actually succeeded. More data points FTR though I'm not sure they'll be useful:
* `flutter run --release` succeeded too
* but OTOH `flutter build ipa --debug` still succeeded, so wouldn't have been a way to detect the issue.

**Comment by gnprice** [comment]
Merging, after adding this:

> Oh I did notice an auto-generated change from `flutter run`:

There was also a similar change in `macos/` after `flutter run` for that. I'd noticed that change yesterday — I was experimenting with the macOS target to try to resolve #1116 and unblock upgrading Firebase — but had assumed it was an existing discrepancy that we just hadn't noticed because we don't use that target as often.

**Pull request merged**
